### PR TITLE
Address test flakiness

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

A mix of fixes that will reduce flakiness caused by to test interactions.
- Ensure batched standalone tests don't race for the same port
- Ensure threads get stopped in teardown - we do our best to identify the threads and cleanly shut them down. For some of them we can't, and we trust they are not doing any harm.

**After many hours, I win.**

----
:books: Documentation preview :books:: https://pytorch-lightning--19022.org.readthedocs.build/en/19022/

cc @carmocca @borda @justusschock @awaelchli

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/strategies/test_ddp_integration.py [hunk 1]
```diff
@@ -202,21 +202,18 @@ def training_step(self, batch, batch_idx):
...
+def test_find_unused_parameters_exception_ddp_spawn():
     """Test that the DDP strategy can change PyTorch's error message so that it's more useful for Lightning users."""
     trainer = Trainer(accelerator="cpu", devices=1, strategy="ddp_spawn", max_steps=2)
     with pytest.raises(
         ProcessRaisedException, match="It looks like your LightningModule has parameters that were not used in"
     ):
         trainer.fit(UnusedParametersModel())
 
+
+@RunIf(min_cuda_gpus=1, standalone=True)
```
Comment (carmocca) [inline tests/tests_pytorch/strategies/test_ddp_integration.py:214]: This test doesn't use CUDA?

```suggestion
@RunIf(standalone=True)
```

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/strategies/test_ddp_integration.py (hunk 1 above)
Comment (awaelchli) [reply tests/tests_pytorch/strategies/test_ddp_integration.py:214]: standalone doesn't run on cpu-only configuration last time I checked?

File: tests/tests_pytorch/strategies/test_ddp_integration.py [hunk 2]
```diff
@@ -202,22 +201,20 @@ def training_step(self, batch, batch_idx):
...
 
-@pytest.mark.skipif(
-    # TODO: investigate threading issue in this configuration
-    _IS_WINDOWS,
-    # and (sys.version_info.major, sys.version_info.minor) == (3, 11)
-    # and compare_version("torch", operator.eq, "2.1.0", use_base_version=True)
-    reason="threading issue",
-)
-def test_find_unused_parameters_exception():
+@RunIf(standalone=True)
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_ddp_integration.py:204]: Couldn't find the "why" to this mystery, but standalone seems to work and doesn't require us to skip.

File: .github/workflows/ci-tests-pytorch.yml [hunk 3]
```diff
@@ -203,7 +203,7 @@ jobs:
         run: |
           echo $GITHUB_RUN_ID
           python -m coverage run --source ${{ env.COVERAGE_SCOPE }} \
-            -m pytest . -v --timeout=60 --durations=50 --random-order-seed=$GITHUB_RUN_ID
+            -m pytest . -v --timeout=60 --durations=50 --random-order-seed=5
```
Comment (awaelchli) [inline .github/workflows/ci-tests-pytorch.yml:206]: todo run a few more, then revert

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/run_standalone_tests.sh [hunk 4]
```diff
@@ -51,7 +51,7 @@ function show_batched_output {
   if [ -f standalone_test_output.txt ]; then  # if exists
     cat standalone_test_output.txt
     # heuristic: stop if there's mentions of errors. this can prevent false negatives when only some of the ranks fail
-    if grep -iE 'error|exception|traceback|failed' standalone_test_output.txt | grep -qvE 'on_exception|xfailed'; then
+    if grep -iE 'error|exception|traceback|failed' standalone_test_output.txt | grep -qvE 'on_exception|xfailed|test_find_unused_parameters_exception'; then
```
Comment (carmocca) [inline tests/tests_pytorch/run_standalone_tests.sh:54]: I would rename the test to `test_find_unused_parameters_raises` which is what we for other similar tests

File: tests/tests_pytorch/strategies/test_ddp_integration.py [hunk 5]
```diff
@@ -202,22 +201,20 @@ def training_step(self, batch, batch_idx):
...
+def test_find_unused_parameters_exception_ddp_spawn():
     """Test that the DDP strategy can change PyTorch's error message so that it's more useful for Lightning users."""
-    trainer = Trainer(accelerator="cpu", devices=1, strategy="ddp_spawn", max_steps=2)
+    trainer = Trainer(accelerator="cpu", devices=2, strategy="ddp_spawn", max_steps=2, logger=False)
```
Comment (carmocca) [inline tests/tests_pytorch/strategies/test_ddp_integration.py:207]: Wouldn't devices=1 help speed it up?

**Review by carmocca — APPROVED** [review]
Nice!

**Pull request merged**